src/phasorpy/_phasorpy.pyx

```diff
@@ -2034,3 +2036,106 @@ def _signal_denoise_vector(
             n = <float_t> (<double> integrated[i] / sum)
             for m in range(samples):
                 denoised[i, m] *= n
+
+
+###############################################################################
+# Filtering functions
+
+cdef float_t _quickselect(
+    float_t *arr,
+    ssize_t left,
+    ssize_t right,
+    const ssize_t k
+) noexcept nogil:
+    """Quickselect algorithm to find the k-th smallest element."""
+    cdef:
+        ssize_t i, pivot_index, pivot_new_index
+        float_t pivot_value, temp
+
+    while left <= right:
+        pivot_index = left + (right - left) // 2
+        pivot_value = arr[pivot_index]
+        temp = arr[pivot_index]
+        arr[pivot_index] = arr[right]
+        arr[right] = temp
+        pivot_new_index = left
+        for i in range(left, right):
+            if arr[i] < pivot_value:
+                temp = arr[i]
+                arr[i] = arr[pivot_new_index]
+                arr[pivot_new_index] = temp
+                pivot_new_index += 1
+        temp = arr[right]
+        arr[right] = arr[pivot_new_index]
+        arr[pivot_new_index] = temp
+
+        if pivot_new_index == k:
+            return arr[k]
+        if pivot_new_index < k:
+            left = pivot_new_index + 1
+        else:
+            right = pivot_new_index - 1
+
+    return arr[k]
+
+
+cdef float_t _median(const float_t *values, const ssize_t n) noexcept nogil:
+    """Calculate the median of an array of values."""
+    if n % 2 == 0:
+        return (
+            (_quickselect(values, 0, n - 1, n // 2 - 1) +
+             _quickselect(values, 0, n - 1, n // 2)) / 2
+        )
+
+    return _quickselect(values, 0, n - 1, n // 2)
+
+
+def _apply_2d_median_filter(
+    const float_t[:, :] image,
+    float_t[:, :] filtered_image,
+    const ssize_t kernel_size
+):
+    """Apply a 2D median filter ignoring NaN."""
+    if kernel_size <= 0:
+        raise ValueError("kernel_size must be greater than 0")
+
+    cdef:
+        int rows = image.shape[0]
+        int cols = image.shape[1]
+        int k = kernel_size // 2
+        int i, j, di, dj
+        int ki, kj
+        int valid_count
```

Comment:
`int` -> `ssize_t`

Reply:
You are right, sorry I forgot to change this.